### ISSUE TITLE
Always enable Querier BatchIterators

### ISSUE THREAD

**Issue by yeya24** [cortexproject/cortex#5814]
## Proposal

`-querier.batch-iterators` was marked as the default option in https://github.com/cortexproject/cortex/pull/2344, which was 4 years ago.

https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L154 

There are other two options `Iterators` and `None`. Both goes to different code paths.

Since Batch Iterator has been used for a long time in various production environment and is pretty stable. I propose that we make Batch Iterator always on and remove other two options.

This could simplify the codebase quite a bit and make it easier to support querying native histograms.

**Comment by alanprot** [comment]
+1

**Comment by yeya24** [comment]
The plan is to mark this flag as deprecated in v1.17.0 and totally remove it in v1.19.0.

**Comment by yeya24** [comment]
Actually I want to verify whether we can remove the chunks iterators Cortex currently maintains 
 or not https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L154 and reuse Prometheus upstream storage interface as much as possible.

If batch iterators proved to have a much better performance, then we should definitely keep it. Otherwise, we can change to use the upstream interface instead. By reusing the upstream interface, there is almost no code change required to support native histograms on the query path.

**Comment by yeya24** [comment]
> Actually I want to verify whether we can remove the chunks iterators Cortex currently maintains
or not https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L154 and reuse Prometheus upstream storage interface as much as possible.

Another benefit of totally removing chunks and Cortex's own iterators is to support Thanos downsampled chunks out of box. Otherwise, we have to implement those iterators for it as well.

**Comment by yeya24** [comment]
https://gist.github.com/yeya24/3393b37230e61299c2b1b4d68731cb9b

I just tested the iterating performance of Cortex Iterator, Batch Iterator and using the `ChainSampleIteratorFromIterables` Iterator from upstream Prometheus code. It looks like the Prometheus iterator has the best performance of all 3.

I proposed that we remove the existing Cortex iterators and switch to use upstream Prometheus iterator.

```
goos: darwin
goarch: arm64
pkg: github.com/cortexproject/cortex/pkg/querier
BenchmarkIterator/batch-10                  5115           7021716 ns/op           87911 B/op       1019 allocs/op
BenchmarkIterator/batch-10                  5181           7014312 ns/op           87903 B/op       1019 allocs/op
BenchmarkIterator/batch-10                  5174           7001015 ns/op           87891 B/op       1019 allocs/op
BenchmarkIterator/batch-10                  5185           7016972 ns/op           87892 B/op       1019 allocs/op
BenchmarkIterator/batch-10                  5188           7014848 ns/op           87889 B/op       1019 allocs/op
BenchmarkIterator/iterators-10              3802           9493879 ns/op          281005 B/op       3032 allocs/op
BenchmarkIterator/iterators-10              3752           9542021 ns/op          281018 B/op       3032 allocs/op
BenchmarkIterator/iterators-10              3778           9494558 ns/op          280996 B/op       3032 allocs/op
BenchmarkIterator/iterators-10              3746           9517363 ns/op          280994 B/op       3032 allocs/op
BenchmarkIterator/iterators-10              3770           9545949 ns/op          280996 B/op       3032 allocs/op
BenchmarkIterator/prom_iterators-10                 8252           4355611 ns/op          195268 B/op       1015 allocs/op
BenchmarkIterator/prom_iterators-10                 8228           4583210 ns/op          195267 B/op       1015 allocs/op
BenchmarkIterator/prom_iterators-10                 8059           4354646 ns/op          195266 B/op       1015 allocs/op
BenchmarkIterator/prom_iterators-10                 8353           4343057 ns/op          195264 B/op       1015 allocs/op
BenchmarkIterator/prom_iterators-10                 8143           4347176 ns/op          195265 B/op       1015 allocs/op
PASS
ok      github.com/cortexproject/cortex/pkg/querier     553.479s
```

**Comment by yeya24** [comment]
I tested the performance of Cortex `batch Iterator`, `Iterator` and upstream Prometheus iterator (mainly storage.ChainSampleIteratorFromIterables) performance and put the benchmark results [here](https://gist.github.com/yeya24/3393b37230e61299c2b1b4d68731cb9b).

The test covers 2 operations. `Next` and `Seek`.

Benchmark result is below.

### Seek
1. When number of chunks is low, Cortex Iterator has the best performance. Next is batch iterator and Prometheus iterator is the worst.
2. When number of chunks are large, Cortex Batch Iterator shows better performance than Cortex Iterator. Prometheus iterator is way worse. Like 20X slower than Cortex Batch Iterator.

### Next
1. If we iterate over all samples in chunks, Prometheus Iterator has the best performance. Cortex Batch Iterator performance is slightly better than Cortex Iterator.
2. When number of chunks increase, Cortex Batch Iterator performance increases. The more chunks we have, the worse performance we get on Prometheus iterator
3. When the number of chunks is large and we stop early when iterating samples, Cortex Batch Iterator has the best performance and Cortex Iterator has the worst performance.

TLDR:
1. I don't think we should replace Cortex Batch Iterator and Cortex Iterator with Prometheus Iterator due to Prometheus Iterator's poor `Seek` performance
2. Since Cortex Batch Iterator has been the default for 4 years, I think we should remove Cortex Iterator even though it seems to have better `Seek` performance at some scenario.
